pkg/expression/builtin_time.go

```diff
@@ -809,6 +811,166 @@ func (b *builtinLastMonthSig) evalString(ctx EvalContext, row chunk.Row) (string
+		date.SetCoreTime(types.FromDate(0, 0, 0, hour, minute, second, resultTime.Nanosecond()/1000))
+		return date, false, nil
+	}
+
+	var result types.Time
+	result.SetCoreTime(types.FromGoTime(resultTime))
+	tc := typeCtx(ctx)
+	overflow, err := types.DateTimeIsOverflow(tc, date)
+	if err := handleInvalidTimeError(ctx, err); err != nil {
+		return types.ZeroTime, true, err
+	}
+	if overflow {
+		return types.ZeroTime, true, handleInvalidTimeError(ctx, types.ErrDatetimeFunctionOverflow.GenWithStackByArgs("datetime"))
+	}
+
+	return result, false, nil
+}
+
+type nextDayFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *nextDayFunctionClass) getFunction(ctx BuildContext, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETDatetime, types.ETDatetime, types.ETString)
+	if err != nil {
+		return nil, err
+	}
+
+	bf.setDecimalAndFlenForDatetime(types.MinFsp)
+	sig := &builtinNextDaySig{bf}
+	return sig, nil
+}
+
+type builtinNextDaySig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinNextDaySig) Clone() builtinFunc {
+	newSig := &builtinNextDaySig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+// evalTime evals a builtinNextDaySig.
+func (b *builtinNextDaySig) evalTime(ctx EvalContext, row chunk.Row) (types.Time, bool, error) {
+	// need a map from string to actual weekdays
+	dayMap := map[string]time.Weekday{
+		"SUN": time.Sunday, "SUNDAY": time.Sunday,
+		"MON": time.Monday, "MONDAY": time.Monday,
+		"TUE": time.Tuesday, "TUESDAY": time.Tuesday,
+		"WED": time.Wednesday, "WEDNESDAY": time.Wednesday,
+		"THU": time.Thursday, "THURSDAY": time.Thursday,
+		"FRI": time.Friday, "FRIDAY": time.Friday,
+		"SAT": time.Saturday, "SATURDAY": time.Saturday,
+	}
+
+	date, isNull, err := b.args[0].EvalTime(ctx, row)
+	if isNull || err != nil {
+		return types.ZeroTime, isNull, handleInvalidTimeError(ctx, err)
+	}
+	weekday, isNull, err := b.args[1].EvalString(ctx, row)
+	if isNull || err != nil {
+		return types.ZeroTime, isNull, handleInvalidTimeError(ctx, err)
+	}
+
+	// deal with case sensitivity
+	weekday = strings.ToUpper(weekday)
+
+	targetWeekday, valid := dayMap[weekday]
+	if !valid {
+		return types.ZeroTime, isNull, errors.Errorf("given weekday %s is not valid", weekday)
+	}
+	currentWeekday := date.Weekday()
+	daysToAdd := (int64(targetWeekday)-int64(currentWeekday)+6)%7 + 1
+
+	//calculate the exact date
+	orgTime, err := date.GoTime(time.UTC)
+	if err != nil {
+		return types.ZeroTime, true, handleInvalidTimeError(ctx, err)
+	}
+
+	resultTime, err := types.AddDate(0, 0, daysToAdd, orgTime)
+	if err != nil {
+		return types.ZeroTime, true, handleInvalidTimeError(ctx, types.ErrDatetimeFunctionOverflow.GenWithStackByArgs("datetime"))
+	}
+
+	// Ensure time is 00:00:00
+	resultTime.Truncate(24 * time.Hour)
+
+	// in case of datetime overflow
+	if resultTime.Year() == 0 {
```

Comment:
If the input date is valid, then the check of `resultTime.Year()` is redundant?

Reply:
the `addData` called above checked for upper bound of `DateTime`, but not the lowerbound. Here is to check and see if it is still above minimum value so that issue like https://github.com/pingcap/tidb/issues/11329 does not happen.